### PR TITLE
Add CRD for jwt-provider config entry

### PR DESCRIPTION
**Pull request by pglass**
Changes proposed in this PR:

Adds a CRD for the jwt-provider config entry which is included in Consul 1.16

How I've tested this PR:

* Manually
* Unit tests

How I expect reviewers to test this PR:

Checklist:
- [x] Tests added
- [x] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -138,7 +138,7 @@ ifeq (, $(shell which controller-gen))
 	CONTROLLER_GEN_TMP_DIR=$$(mktemp -d) ;\
 	cd $$CONTROLLER_GEN_TMP_DIR ;\
 	go mod init tmp ;\
-	go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.8.0 ;\
+	go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.10.0 ;\
```
Comment (pglass) [inline Makefile:141]: Without v0.10.0, I got a diff at some point while adding this CRD

File: acceptance/tests/config-entries/config_entries_test.go [hunk 2]
```diff
@@ -178,6 +178,22 @@ func TestController(t *testing.T) {
 					require.Equal(r, "certFile", terminatingGatewayEntry.Services[0].CertFile)
 					require.Equal(r, "keyFile", terminatingGatewayEntry.Services[0].KeyFile)
 					require.Equal(r, "sni", terminatingGatewayEntry.Services[0].SNI)
+
+					// jwt-provider
+					entry, _, err = consulClient.ConfigEntries().Get(api.JWTProvider, "test-jwt-provider", nil)
```
Comment (pglass) [inline acceptance/tests/config-entries/config_entries_test.go:183]: Right now, I can't get this test to run successfully locally,

```
    config_entries_test.go:73: 2023-05-30T12:42:29-05:00 Waiting for pods with label "release=test-2vzrrv" to be ready.
    config_entries_test.go:73: 2023-05-30T12:43:05-05:00 Finished waiting for pods to be ready.
    tunnel.go:153: 2023-05-30T12:43:05-05:00 Creating a port forwarding tunnel for resource pod/test-2vzrrv-consul-server-0 routing local port 49385 to remote port 8500
TestController/secure:_false,_vault:_false 2023-05-30T12:43:05-05:00 client.go:33: Configuring kubectl using config file /Users/pglass/.kube/config with context kind-dc1
    tunnel.go:185: 2023-05-30T12:43:05-05:00 Selected pod test-2vzrrv-consul-server-0 to open port forward to
    tunnel.go:198: 2023-05-30T12:43:05-05:00 Using URL https://127.0.0.1:65483/api/v1/namespaces/default/pods/test-2vzrrv-consul-server-0/portforward to create portforward
    tunnel.go:247: 2023-05-30T12:43:05-05:00 Successfully created port forwarding tunnel
    config_entries_test.go:82: 2023-05-30T12:43:05-05:00 creating custom resources
    command.go:100: 2023-05-30T12:43:05-05:00 Running command kubectl with args [--context kind-dc1 --namespace default apply -k ../fixtures/bases/crds-oss]
    command.go:179: 2023-05-30T12:43:06-05:00 ingressgateway.consul.hashicorp.com/ingress-gateway created
    command.go:179: 2023-05-30T12:43:06-05:00 mesh.consul.hashicorp.com/mesh created
    command.go:179: 2023-05-30T12:43:06-05:00 proxydefaults.consul.hashicorp.com/global created
    command.go:179: 2023-05-30T12:43:06-05:00 servicedefaults.consul.hashicorp.com/defaults created
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-servicedefaults.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-servicedefaults?timeout=10s": EOF
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-servicedefaults.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-servicedefaults?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-servicedefaults.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-servicedefaults?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-servicedefaults.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-servicedefaults?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-servicedefaults.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-servicedefaults?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-servicedefaults.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-servicedefaults?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-serviceintentions.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-serviceintentions?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-serviceresolver.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-serviceresolver?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-servicerouter.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-servicerouter?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-servicesplitter.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-servicesplitter?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:179: 2023-05-30T12:43:07-05:00 Error from server (InternalError): error when creating "../fixtures/bases/crds-oss": Internal error occurred: failed calling webhook "mutate-terminatinggateway.consul.hashicorp.com": failed to call webhook: Post "https://test-2vzrrv-consul-connect-injector.default.svc:443/mutate-v1alpha1-terminatinggateway?timeout=10s": dial tcp 10.96.244.94:443: connect: connection refused
    command.go:100: 2023-05-30T12:43:07-05:00 Running command kubectl with args [--context kind-dc1 --namespace default apply -k ../fixtures/bases/crds-oss]
```

File: charts/consul/templates/connect-inject-clusterrole.yaml [hunk 3]
```diff
@@ -31,6 +31,7 @@ rules:
   - peeringacceptors
   - peeringdialers
   {{- end }}
+  - jwtproviders
```
Comment (pglass) [inline charts/consul/templates/connect-inject-clusterrole.yaml:34]: I think I got the plurality right here, but there were so many places to update I'm not positive.

File: control-plane/api/v1alpha1/jwtprovider_types.go [hunk 4]
```diff
@@ -0,0 +1,663 @@
...
+		return nil
+	}
+	return &capi.JWTLocation{
+		Header:     j.Header.toConsul(),
+		QueryParam: j.QueryParam.toConsul(),
+		Cookie:     j.Cookie.toConsul(),
+	}
+}
+
+func (j *JWTLocation) validate(path *field.Path) field.ErrorList {
```
Comment (pglass) [inline control-plane/api/v1alpha1/jwtprovider_types.go:145]: Throughout this file, I tried to match the validations we implemented in https://github.com/hashicorp/consul/blob/main/agent/structs/config_entry_jwt_provider.go

File: control-plane/api/v1alpha1/jwtprovider_webhook.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	client.Client
+}
+
+// NOTE: The path value in the below line is the path to the webhook.
+// If it is updated, run code-gen, update subcommand/controller/command.go
+// and the consul-helm value for the path to the webhook.
+//
+// NOTE: The below line cannot be combined with any other comment. If it is it will break the code generation.
+//
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-v1alpha1-jwtproviders,mutating=true,failurePolicy=fail,groups=consul.hashicorp.com,resources=jwtproviders,versions=v1alpha1,name=mutate-jwtprovider.consul.hashicorp.com,sideEffects=None,admissionReviewVersions=v1beta1;v1
```
Comment (pglass) [inline control-plane/api/v1alpha1/jwtprovider_webhook.go:34]: Carefully compare this to the webhook file for another type to make sure it looks right

File: control-plane/api/v1alpha1/jwtprovider_webhook.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	client.Client
+}
+
+// NOTE: The path value in the below line is the path to the webhook.
+// If it is updated, run code-gen, update subcommand/controller/command.go
+// and the consul-helm value for the path to the webhook.
+//
+// NOTE: The below line cannot be combined with any other comment. If it is it will break the code generation.
+//
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-v1alpha1-jwtprovider,mutating=true,failurePolicy=fail,groups=consul.hashicorp.com,resources=jwtproviders,versions=v1alpha1,name=mutate-jwtprovider.consul.hashicorp.com,sideEffects=None,admissionReviewVersions=v1beta1;v1
```
Comment (roncodingenthusiast) [inline control-plane/api/v1alpha1/jwtprovider_webhook.go:34]:
```suggestion
// +kubebuilder:webhook:verbs=create;update,path=/mutate-v1alpha1-jwtproviders,mutating=true,failurePolicy=fail,groups=consul.hashicorp.com,resources=jwtproviders,versions=v1alpha1,name=mutate-jwtprovider.consul.hashicorp.com,sideEffects=None,admissionReviewVersions=v1beta1;v1
```

Wondering if this should be plural: (samenessgroup and others have this as plural)

File: control-plane/config/webhook/manifests.yaml [hunk 7]
```diff
@@ -50,6 +50,27 @@ webhooks:
...
     - ingressgateways
   sideEffects: None
+- admissionReviewVersions:
+  - v1beta1
+  - v1
+  clientConfig:
+    service:
+      name: webhook-service
+      namespace: system
+      path: /mutate-v1alpha1-jwtprovider
```
Comment (roncodingenthusiast) [inline control-plane/config/webhook/manifests.yaml:60]: other examples in this repo seem to have this as plural. I am sure it doesn't matter but if you take the code suggestion above then you might need to update this too

**Review by roncodingenthusiast — APPROVED** [review]
🎉  This looks good, a few non-blocking neats to consider

File: control-plane/api/v1alpha1/jwtprovider_webhook.go (hunk 6 above)
Comment (pglass) [reply control-plane/api/v1alpha1/jwtprovider_webhook.go:34]: I'm still wondering...

Ingress gateway is not plural:

https://github.com/hashicorp/consul-k8s/blob/2261a1b5aa4d44516e7b5722527d59d4c23875f5/control-plane/api/v1alpha1/ingressgateway_webhook.go#L34

Mesh config entry is not plural:

https://github.com/hashicorp/consul-k8s/blob/2261a1b5aa4d44516e7b5722527d59d4c23875f5/control-plane/api/v1alpha1/mesh_webhook.go

Service resolver is not plural:

https://github.com/hashicorp/consul-k8s/blob/2261a1b5aa4d44516e7b5722527d59d4c23875f5/control-plane/api/v1alpha1/serviceresolver_webhook.go#L34

Service router is not plural:

https://github.com/hashicorp/consul-k8s/blob/2261a1b5aa4d44516e7b5722527d59d4c23875f5/control-plane/api/v1alpha1/servicerouter_webhook.go#L34

Things like proxy-defaults and service-defaults and service-intentions are plural, but those config entry names were already plural where as jwt-provider is singular.

File: control-plane/config/webhook/manifests.yaml (hunk 7 above)
Comment (pglass) [reply control-plane/config/webhook/manifests.yaml:60]: These are generated from the kubebuilder annotations, and some are plural but many aren't (ingress gateway, service-router, service-resolver, mesh, etc).

**Pull request merged**